lib/ret/hub.ex

```diff
@@ -897,6 +897,57 @@ end
 
 defimpl Canada.Can, for: Ret.Account do
   alias Ret.{Hub, AppConfig}
+  alias Ret.Api.{Scopes, Credentials}
+
+  # Non-admin accounts can create their own account tokens
+  def can?(%Ret.Account{account_id: account_id}, :create_credentials, %{
+        subject_type: :account,
+        scopes: scopes,
+        account_id: account_id
+      }) do
+    # TODO: Is it correct to disallow the create_accounts scope for account type tokens?
+    # If so, why? Do we need to enforce this everywhere?
+    Scopes.create_accounts() not in scopes
+  end
+
+  # Admin accounts can create tokens on behalf of other users
+  # TODO: Should we even allow this?
+  def can?(%Ret.Account{is_admin: true}, :create_credentials, %{subject_type: :account, scopes: scopes}) do
+    Scopes.create_accounts() not in scopes
+  end
+
+  def can?(%Ret.Account{is_admin: is_admin}, :create_credentials, %{subject_type: :app}) do
+    is_admin
+  end
+
+  def can?(%Ret.Account{}, :create_credentials, _params) do
+    false
+  end
+
+  def can?(%Ret.Account{is_admin: is_admin}, :list_credentials, :app) do
+    is_admin
+  end
+
+  def can?(%Ret.Account{}, :list_credentials, :account) do
+    # TODO: Allow admins to disable this in config
+    true
+  end
+
+  def can?(%Ret.Account{}, :list_credentials, _subject_type) do
+    false
+  end
+
+  def can?(%Ret.Account{account_id: account_id}, :revoke_credentials, %Credentials{account_id: account_id}) do
+    true
+  end
+
+  def can?(%Ret.Account{is_admin: true}, :revoke_credentials, %Credentials{}) do
+    true
+  end
+
+  def can?(%Ret.Account{}, :revoke_credentials, %Credentials{}) do
+    false
+  end
```

Comment:
I agree, but I don't think I can break the `Canada` `impl` for `Account`s across two files. (I tried doing this for `Atom` but didn't have success, but I might have done something wrong.) Instead I put these `Account` checks for credentials-related actions here. 

I put the `impl` for `Credentials` into its own file (`can_credentials.ex`) because I knew eventually we would have actions that did not have to do with rooms. We could do the same for accounts (e.g. `can_account.ex`). 

All the `Canada` code is sorted by `subject` due to it being a protocol (which matches on the first argument), but it might also make sense to sort functions like these by `action`. Not sure what's better.

[  def can?(subject, action, resource)](https://github.com/jarednorman/canada/blob/b21d17f7b4561aca8461614106595054d4bfd557/lib/canada/can.ex#L5)

Reply:
Ah, right. I think moving everything into a `can_accounts.ex` is a good idea.